rfd/headless-login.md

```diff
@@ -0,0 +1,367 @@
+
+This endpoint is used by Teleport clients to update headless login request state to approved or denied. If the client requests to approve, the client will receive an MFA challenge and will need to reply with a valid MFA challenge response. To support this multi-step request, the rpc provides an rpc stream to both sides of the connection.
+
+The endpoint is only authorized for the user who requested Headless authentication (and for server roles).
+
+```proto
+service AuthService {
+  rpc UpdateHeadlessLoginRequestState(stream UpdateHeadlessLoginRequestStateRequest) returns (stream UpdateHeadlessLoginRequestStateResponse);
+}
+
+message UpdateHeadlessLoginRequestStateRequest {
+  oneof Request {
+    UpdateHeadlessLoginRequestStateRequestInit Init = 1;
+    MFAAuthenticateResponse MFAResponse = 2;
+  }
+}
+
+message UpdateHeadlessLoginRequestStateRequestInit {
+  string RequestID = 1;
+  State NewState = 2;
+}
+
+enum State {
+  PENDING = 0;
+  DENIED = 1;
+  APPOVED = 2;
+}
+
+// MFAAuthenticateResponse is a response to MFAAuthenticateChallenge using one
+// of the MFA devices registered for a user.
+message MFAAuthenticateResponse {
+  oneof Response {
+    // Removed: U2FResponse U2F = 1;
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
+  }
+}
+
+
+message UpdateHeadlessLoginRequestStateResponse {
+  MFAAuthenticateChallenge MFAChallenge = 1;
+}
+
+// MFAAuthenticateChallenge is a challenge for all MFA devices registered for a
+// user.
+message MFAAuthenticateChallenge {
+  reserved 1; // repeated U2FChallenge U2F
+  // TOTP is a challenge for all TOTP devices registered for a user. When
+  // this field is set, any TOTP device a user has registered can be used to
+  // respond.
+  TOTPChallenge TOTP = 2;
+  // WebauthnChallenge contains a Webauthn credential assertion used for
+  // login/authentication ceremonies.
+  // Credential assertions hold, among other information, a list of allowed
+  // credentials for the ceremony (one for each U2F or Webauthn device
+  // registered by the user).
+  webauthn.CredentialAssertion WebauthnChallenge = 3;
+}
+```
+
+#### `POST /webapi/login/headless/finish`
+
+This endpoint is used to initiate headless login and get a request ID. Like other login endpoints, this endpoint is not authenticated and can be called by anyone with access to the Teleport Proxy address. However, the request will be denied the the request ID, user, and public key do not match the `HeadlessLoginRequest` saved on the auth server.
+
+```go
+type FinishHeadlessLoginRequest struct {
+    // RequestID is the headless login request ID
+    RequestID string `json:"request_id"`
+    // User is a teleport username
+    User string `json:"user"`
+    // PubKey is a public key user wishes to sign
+    PubKey []byte `json:"pub_key"`
+}
+
+// SSHLoginResponse is a response returned by web proxy, it preserves backwards compatibility
+// on the wire, which is the primary reason for non-matching json tags
```

Comment:
I'm not sure about the compatibility you're referring to here.

Reply:
This was just copy and pasted from our code, I plan on polishing the request/response comments later.